File: token-validation-jaxrs/src/test/java/no/nav/security/token/support/jaxrs/ServerFilterProtectedMethodTest.java
package no.nav.security.token.support.jaxrs;

import jakarta.ws.rs.client.ClientBuilder;
import jakarta.ws.rs.client.Invocation;
import jakarta.ws.rs.core.Response;
import no.nav.security.token.support.core.JwtTokenConstants;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.core.Is.is;

@ActiveProfiles("protected")
@DirtiesContext
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = Config.class)
class ServerFilterProtectedMethodTest {

    @LocalServerPort
    private int port;

    private Invocation.Builder requestWithValidToken(String path) {
        return ClientBuilder.newClient().target("http://localhost:" + port)
                .path(path)
                .request()
                .header(JwtTokenConstants.AUTHORIZATION_HEADER,
                        "Bearer " + JwtTokenGenerator.createSignedJWT("12345678911").serialize());
    }

    private Invocation.Builder requestWithoutToken(String path) {
        return ClientBuilder.newClient().target("http://localhost:" + port)
                .path(path)
                .request();
    }

    @Test
     void that_unprotected_returns_ok_with_valid_token() {
        Response response = requestWithValidToken("unprotected").get();
        assertThat(response.getStatus(), is(equalTo(200)));
    }

    @Test
     void that_protected_returns_200_with_valid_token() {
        Response response = requestWithValidToken("protected").get();
        assertThat(response.getStatus(), is(equalTo(200)));
    }

    @Test
     void that_protected_with_claims_returns_200_with_valid_token() {
        Response response = requestWithValidToken("protected/with/claims").get();
        assertThat(response.getStatus(), is(equalTo(200)));
    }

    @Test
     void that_unprotected_returns_200_without_token() {
        Response response = requestWithoutToken("unprotected").get();
        assertThat(response.getStatus(), is(equalTo(200)));
    }

    @Test
     void that_protected_returns_401_without_token() {
        Response response = requestWithoutToken("protected").get();
        assertThat(response.getStatus(), is(equalTo(401)));
    }

    @Test
     void that_protected_with_claims_returns_401_without_token() {
        Response response = requestWithoutToken("protected/with/claims").get();
        assertThat(response.getStatus(), is(equalTo(401)));
    }

    @Test
     void that_protected_with_claims_returns_403_with_invalid_claims() {
        Response response = requestWithValidToken("protected/with/claims/unknown").get();
        assertThat(response.getStatus(), is(equalTo(403)));
    }

}

File: token-validation-jaxrs/src/test/kotlin/no/nav/security/token/support/jaxrs/ServerFilterProtectedMethodTest.kt
package no.nav.security.token.support.jaxrs

import jakarta.ws.rs.client.ClientBuilder.newClient
import no.nav.security.token.support.core.JwtTokenConstants.AUTHORIZATION_HEADER
import no.nav.security.token.support.jaxrs.JwtTokenGenerator.createSignedJWT
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.http.HttpStatus.FORBIDDEN
import org.springframework.http.HttpStatus.OK
import org.springframework.http.HttpStatus.UNAUTHORIZED
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ActiveProfiles

@ActiveProfiles("protected")
@DirtiesContext
@SpringBootTest(webEnvironment = RANDOM_PORT, classes = [Config::class])
internal class ServerFilterProtectedMethodTest {

    @LocalServerPort
    private val port = 0

    private fun requestWithValidToken(path : String) =
        newClient().target("http://localhost:$port")
            .path(path)
            .request()
            .header(AUTHORIZATION_HEADER, "Bearer " + createSignedJWT("12345678911").serialize())
    private fun requestWithoutToken(path : String) = newClient().target("http://localhost:$port").path(path).request()
    @Test
    fun that_unprotected_returns_ok_with_valid_token() {
        assertEquals(OK.value(), requestWithValidToken("unprotected").get().status)
    }

    @Test
    fun that_protected_returns_200_with_valid_token() {
        assertEquals(OK.value(), requestWithValidToken("protected").get().status)
    }

    @Test
    fun that_protected_with_claims_returns_200_with_valid_token() {
        assertEquals(OK.value(), requestWithValidToken("protected/with/claims").get().status)
    }

    @Test
    fun that_unprotected_returns_200_without_token() {
        assertEquals(OK.value(), requestWithoutToken("unprotected").get().status)
    }

    @Test
    fun that_protected_returns_401_without_token() {
        assertEquals(UNAUTHORIZED.value(), requestWithoutToken("protected").get().status)
    }

    @Test
    fun that_protected_with_claims_returns_401_without_token() {
        assertEquals(UNAUTHORIZED.value(), requestWithoutToken("protected/with/claims").get().status)
    }

    @Test
    fun that_protected_with_claims_returns_403_with_invalid_claims() {
        assertEquals(FORBIDDEN.value(), requestWithValidToken("protected/with/claims/unknown").get().status)
    }
}